pkg/cloudproviders/k8s/k8s.go

```diff
@@ -0,0 +1,380 @@
+	a := util.GetSubViper(options.Viper, "k8s")
+	a.SetDefault(ParamKubeconfigContext, DefaultKubeconfigContext)
+	a.SetDefault(ParamKubeconfigPath, DefaultKubeconfigPath)
+
+	// Set up the k8s client
+	var restConfig *rest.Config
+	var err error
+	kubeconfigPath := a.GetString(ParamKubeconfigPath)
+	kubeconfigContext := a.GetString(ParamKubeconfigContext)
+	if kubeconfigPath != "" {
+		configOverrides := &clientcmd.ConfigOverrides{ClusterInfo: clientcmdapi.Cluster{Server: ""}}
+		if kubeconfigContext != "" {
+			configOverrides.CurrentContext = kubeconfigContext
+		}
+
+		restConfig, err = clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+			&clientcmd.ClientConfigLoadingRules{ExplicitPath: kubeconfigPath},
+			configOverrides).ClientConfig()
+	} else {
+		restConfig, err = rest.InClusterConfig()
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	apiQPS := a.GetFloat64(ParamAPIQPS)
+	apiQPSBurstFactor := a.GetFloat64(ParamAPIQPSBurstFactor)
+	restConfig.RateLimiter = flowcontrol.NewTokenBucketRateLimiter(float32(apiQPS), int(apiQPS*apiQPSBurstFactor))
+	restConfig.UserAgent = UserAgent + "/" + options.Version
+
+	clientset, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return nil, err
+	}
+	return NewProvider(options, clientset)
+}
+
+// NewProvider returns a new k8s provider with the given k8s client configuration.
+func NewProvider(options gostatsd.CloudProviderOptions, clientset kubernetes.Interface) (gostatsd.CloudProvider, error) {
+	a := util.GetSubViper(options.Viper, "k8s")
+	a.SetDefault(ParamAPIQPS, DefaultAPIQPS)
+	a.SetDefault(ParamAPIQPSBurstFactor, DefaultAPIQPSBurstFactor)
+	a.SetDefault(ParamAnnotationTagWhitelist, DefaultAnnotationTagWhitelist)
+	a.SetDefault(ParamLabelTagWhitelist, DefaultLabelTagWhitelist)
+	a.SetDefault(ParamResyncPeriod, DefaultResyncPeriod)
+	a.SetDefault(ParamWatchCluster, DefaultWatchCluster)
+
+	// Create the tag whitelists
+	annotationWhitelist, err := stringListToRegexList(a.GetStringSlice(ParamAnnotationTagWhitelist))
+	if err != nil {
+		return nil, err
+	}
+	labelWhitelist, err := stringListToRegexList(a.GetStringSlice(ParamLabelTagWhitelist))
+	if err != nil {
+		return nil, err
+	}
+
+	// Set up informers
+	indexers := cache.Indexers{
+		PodsByIPIndexName: podByIpIndexFunc,
+	}
+
+	watchCluster := a.GetBool(ParamWatchCluster)
+	customWatchOptions := func(*meta_v1.ListOptions) {}
+	if !watchCluster {
+		if options.NodeName == "" {
+			return nil, fmt.Errorf(
+				"'%s' set to false, and node name could not be found from environment variable '%s'",
+				ParamWatchCluster, NodeNameEnvVar)
+		}
+		customWatchOptions = func(lo *meta_v1.ListOptions) {
+			lo.FieldSelector = fmt.Sprintf("spec.nodeName=%s", options.NodeName)
+		}
+	}
+
+	podsInf := core_v1inf.NewFilteredPodInformer(
+		clientset,
+		meta_v1.NamespaceAll,
+		a.GetDuration(ParamResyncPeriod),
+		indexers,
+		customWatchOptions)
+
+	// TODO: we should emit prometheus metrics to fit in with the k8s ecosystem
```

Comment:
Can statser use a separate/different backend than the main pipeline? That would be ideal, I guess.

Reply:
Yep, currently there's only null, internal (sends via the pipeline, but has a different set of configurable tags), and logger implemented, but it would be fairly trivial to add a prom statser I expect, since you're dealing with pre-aggregated data (an actual prom backend would be more complicated because it's aggregated).

The only down side is the configuration around all that is still very minimal (it's in the global config namespace, there's no facility for multiple statsers, etc)

Reply:
I agree with what you're saying here. To clarify, the context of this TODO is just that to fit into the k8s ecosystem properly we should emit Prometheus metrics. I think that the approach you mention is the way to do it, which is why I haven't tackled it here as it's a bit of a big job.

Basically it would just be adding a prometheus backend/statser and then exposing that to the cluster so that k8s operators can monitor it the same way as everything else in the ecosystem 🙂

Reply:
I raised a pair of issues, for backend and internal statser :)